File: count_construct_word/python_impl/count_construct_word.py
from functools import lru_cache


@lru_cache(maxsize=10000)
def count_construct_word(target, word_bank):
    if target == "":
        return 1
    total_count = 0
    for word in word_bank:
        if word in target:
            # print(word)
            if(target.index(word) == 0):
                # suffix = target.split(word)[1]
                suffix = target[len(word):]
                # print("suffix"+suffix)
                number_of_ways = count_construct_word(suffix, word_bank)
                total_count += number_of_ways

    return total_count


print(count_construct_word("purple", tuple(
    ["purp", "p", "ur", "le", "purpl"])))
print(count_construct_word("abcdef", tuple(
    ["ab", "abc", "cd", "def", "abcd"])))
print(count_construct_word("skateboard", tuple([
      "bo", "rd", "ate", "t", "ska", "sk", "boar"])))
print(count_construct_word("enterapotentpot", tuple([
    "a", "p", "en", "enter", "ot", "o", "t"])))
print(count_construct_word("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
                           tuple(["e", "ee", "eee", "eeee", "eeeee", "eeeeee"])))

File: count_construct_word/js_impl/countConstructWord.js
const countConstruct = (target, wordBank) => {
    if (target === "") {
        return 1
    }
    let total = 0;
    for (let word of wordBank) {
        if (target.indexOf(word) === 0) {
            const suffix = target.slice(word.length);
            const numWays = countConstruct(suffix, wordBank);
            total += numWays;

        }
    }
    return total;
}


const memoCountConstruct = (target, wordBank, memo = {}) => {
    if (target in memo) return memo[target];
    if (target === "") {
        return 1
    }
    let total = 0;
    for (let word of wordBank) {
        if (target.indexOf(word) === 0) {
            const suffix = target.slice(word.length);
            const numWays = memoCountConstruct(suffix, wordBank, memo);
            total += numWays;

        }
    }
    memo[target] = total;
    return total;
}




console.log(countConstruct("purple", ["purp", "p", "ur", "le", "purpl"]));
console.log(countConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"]));
console.log(countConstruct("skateboard", [
    "bo", "rd", "ate", "t", "ska", "sk", "boar"
]));

console.log(countConstruct("enterapotentpot", [
    "a", "p", "en", "enter", "ot", "o", "t"
]))
// console.log(countConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
//     ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"]))

console.log("-------------------memoization-----------")

console.log(memoCountConstruct("purple", ["purp", "p", "ur", "le", "purpl"]));
console.log(memoCountConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"]));
console.log(memoCountConstruct("skateboard", [
    "bo", "rd", "ate", "t", "ska", "sk", "boar"
]));

console.log(memoCountConstruct("enterapotentpot", [
    "a", "p", "en", "enter", "ot", "o", "t"
]));
console.log(memoCountConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
    ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"]));